### PR TITLE
add support to DistilBertLMHeadModel

### LINKED ISSUE #14737: Add support to DistilBertLMHeadModel
# 🚀 Feature request

## Background
Hi, I am the first author of the paper ["**TSDAE**: Using Transformer-based Sequential Denoising Auto-Encoder for Unsupervised Sentence Embedding Learning"](https://arxiv.org/abs/2104.06979). The work proposed a new method, TSDAE to train unsupervised-sentence-embedding models with denoising **autoencoder** loss on top of BERT-like PLMs. The code of TSDAE has been integrated into the SBERT repo and one can find the source file [here](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/DenoisingAutoEncoderLoss.py) and the example training script [here](https://github.com/UKPLab/sentence-transformers/tree/master/examples/unsupervised_learning/TSDAE).

## What is missing
As one can tell from the "autoencoder" in the name, TSDAE needs a decoder (i.e. `LMHead`) to work. However, the popular [**DistilBERT**](https://github.com/huggingface/transformers/blob/master/src/transformers/models/distilbert/modeling_distilbert.py) in the current HF-transformers **does not support it as a decoder** (i.e. missing something called `DistilBertLMHeadModel`). 

Previously, I **used to bypass** this issue with [my personal extension](https://gist.github.com/kwang2049/1f0e1f0ce119456284c0af048ba097a7) and one can find the [discussion/issues](https://github.com/UKPLab/sentence-transformers/issues/962) in the SBERT repo about this. Recently, as users reported, I have just found **this way is not really maintainable and continuable**: HF-transformers keeps updating the code all the time and I need to update the extension accordingly then.

## Goal
Add support to **DistilBertLMHeadModel**. Actually, **I have made the changes in my fork** (for PyTorch) and it just needs to be merged (of course, after the code review from the HF community).

## Motivation
With support to DistilBertLMHeadModel, all the decoding functions will be possible for all the amazing distilbert models, e.g. the TSDAE learning, seq2seq tasks, etc.

## Your contribution
I have forked the latest master branch and added the corresponding support for PyTorch. One can check the pull-request for details. Really appreciate it if this issue can be solved.

### PR DESCRIPTION
**Pull request by kwang2049**
# What does this PR do?

Fixes #14737 
Goal: Add support to DistilBertLMHeadModel
Changes: 

1. Modified [modeling_distilbert.py](https://github.com/huggingface/transformers/blob/master/src/transformers/models/distilbert/modeling_distilbert.py): Changed `MultiHeadSelfAttention`, `Transformer`, `TransformerBlock` to model cross-attention and accept arguments `encoder_hidden_states` & `encoder_attention_mask `; and added the new class `DistilBertLMHeadModel`;
2. Expose the new class and register it everywhere needed.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@VictorSanh @thomwolf @patil-suraj

### CODE REVIEW

**Comment by patil-suraj** [comment]
Thanks for the PR! There's a similar PR here #11085, which is almost complete, so let's wait for the author to see if he's still interested to finish it :)

**Comment by kwang2049** [comment]
> Thanks for the PR! There's a similar PR here #11085, which is almost complete, so let's wait for the author to see if he's still interested to finish it :)

Thanks for your attention! OK, let me also have a look at what is going on there and anything I can help:).
